lib/S3Plugin.js

```diff
@@ -0,0 +1,278 @@
+/*
+ * Kuzzle, a backend software, self-hostable and ready to use
+ * to power modern apps
+ *
+ * Copyright 2015-2019 Kuzzle
+ * mailto: support AT kuzzle.io
+ * website: http://kuzzle.io
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+const 
+  AWS = require('aws-sdk'),
+  uuid = require('uuid');
+
+function getProperty(document, path) {
+  if (!document) {
+    return document;
+  }
+
+  const names = path.split('.');
+
+  if (names.length === 1) {
+    return document[names[0]];
+  }
+
+  return getProperty(document[names[0]], names.slice(1).join('.'));
+}
+  
+class S3Plugin {
+  constructor () {
+    this.context = null;
+
+    this.defaultConfig = {
+      bucketName: 'your-s3-bucket',
+      region: 'eu-west-3',
+      signedUrlTTL: 20 * 60 * 1000,
+      redisPrefix: 's3Plugin/fileController'
+    };
+
+    this.hooks = {};
+
+    this.pipes = {};
+
+    this.controllers = {
+      file: {
+        getUploadUrl: 'getUploadUrl',
+        deleteFile: 'deleteFile',
+        getUrl: 'getUrl',
+        validateUpload: 'validateUpload'
+      }
+    };
+
+    this.routes = [
+      { verb: 'get', url: '/getUploadUrl', controller: 'file', action: 'getUploadUrl' },
+      { verb: 'get', url: '/files/:fileKey', controller: 'file', action: 'getUrl' },
+      { verb: 'put', url: '/files/:fileKey/validate', controller: 'file', action: 'validateUpload' },
+      { verb: 'delete', url: '/files/:fileKey', controller: 'file', action: 'deleteFile' }
+    ];
+  }
+
+  init (customConfig, context) {
+    this.config = { ...this.defaultConfig, ...customConfig };
+
+    this.context = context;
+
+    this.s3 = null;
+
+    this.baseFileUrl = `https://s3.${this.config.region}.amazonaws.com/${this.config.bucketName}`;
+
+    this.unavailableMessage = 'File upload unavailable.';
+
+    const { AWS_ACCESS_KEY_ID, AWS_SECRET_ACCESS_KEY, NODE_ENV } = process.env;
+
+    if (NODE_ENV !== 'production') {
+      this.unavailableMessage = `${this.unavailableMessage} AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY are not set.`;
```

Comment:
There is no point starting the plugin if there is no access keys set: this plugin init function should throw, forcing Kuzzle to stop, if these variables aren't set.

Consider this use case: you have a cluster of Kuzzle ndes and, for whatever reason, you start another node but you forgot to deploy the configuration files => you get an inconsistent cluster.